### PR TITLE
Heron unit testing

### PR DESCRIPTION
**Pull request by caleb-sitton-inl**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Not Applicable

##### What are the significant changes in functionality due to this change request?
This implements a complete set of nine HERON unit tests. Specifically, the tests focus on the `src/heron/` `create_componentsets_in_HERON()` function. The `tests/unit_tests/test_heron.py` file contains the tests, which use the `unittest` module to run. This file can be run directly and all unit tests will run. This set of tests is designed to cover the vast majority of potential errors in the function without the length that would be required for a strictly comprehensive test suite.

The `tests/unit_tests/tests` file contains the necessary content for the unit tests to be run through rook by running `FORCE/run_tests`. Rook does not have a significant role in the testing itself; each test is run as a "generic executable" using the `unittest` module, which reports back a failure, success, or error. Output is handled by, and thus formatted by, rook. An advantage to this system is that rook will run both unit and integration tests with `run_tests`.

The `tests/unit_tests/README.md` file contains a description of what each individual unit test is intended to check. 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 3. Automated Tests should pass, including run_tests, pylint, and manual building tests.
- [ ] 4. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test.
- [ ] 5. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 6. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.

### CODE REVIEW

File: tests/unit_tests/test_heron.py [hunk 1]
```diff
@@ -1,61 +1,669 @@
...
+          self.assertEqual(len(cashflow_capex), 1)
+          ref_price = cashflow_capex[0].findall('./reference_price/fixed_value')
+          self.assertEqual(ref_price[0].text, '-2200')
+
+      elif comp.get('name') == 'Component2':
+        economics = comp.findall('./economics')
+
+        # Verify ProjectTime was not corrupted
+        with self.subTest("Non-cashflow child node of economics has been corrupted"):
+          proj_time = economics[0].findall('./ProjectTime')
```
Comment (GabrielSoto-INL) [inline tests/unit_tests/test_heron.py:315]: this should also be `lifetime`

File: tests/unit_tests/test_heron.py [hunk 2]
```diff
@@ -1,61 +1,669 @@
...
+                              <economics>
+                                <ProjectTime>1</ProjectTime>
+                                <CashFlow name="capex_Comp2" npv_exempt="True">
+                                  <reference_driver>
+                                    <fixed_value>1234</fixed_value>
+                                  </reference_driver>
+                                  <reference_price>
+                                    <ROM rom_stuff="stuff">more_rom_stuff</ROM>
+                                  </reference_price>
+                                  <scaling_factor_x>
```
Comment (GabrielSoto-INL) [inline tests/unit_tests/test_heron.py:190]: I think it's good to check these mixed-ValuedParam cases. From what I'm seeing here, every child node here gets replaced with a value from the `mock_open` callable below. Now this might be a question of ignorance, but: would it be possible to leave one of these parameters unchanged? By that I mean, for example, that the end goal for `Component 2` is to replace the `reference_driver`, the `reference_price`, and the `driver` with values read in from Line 246-248, BUT we wanted to keep the `scaling_factor_x` as an uncertainty. Is that possible to test for/a current capability of the method?

File: tests/unit_tests/test_heron.py [hunk 3]
```diff
@@ -1,61 +1,669 @@
...
+                          "Reference Price (USD)": 2000,
+                          "Scaling Factor": 0.5
+                      }""")
+  def test_empty_compsets_folder(self, mock_open, mock_listdir, mock_parse):
+    # Set up the parse mock to return an XML tree
+    mock_parse.return_value = self.tree
+    mock_listdir.return_value = []
+
+    # Call the function
+    result_tree = create_componentsets_in_HERON("/fake/folder", "/fake/heron_input.xml")
```
Comment (GabrielSoto-INL) [inline tests/unit_tests/test_heron.py:566]: this might be out of the scope of this PR, but should this scenario cause an error or at least a warning?

File: tests/unit_tests/test_heron.py [hunk 4]
```diff
@@ -1,61 +1,669 @@
...
-        ref_driver = cashflows[0].find('./reference_driver/fixed_value')
-        self.assertEqual(ref_driver.text, '1.0')  # The driver should have been converted from kW to MW
+    # Call the function
+    result_tree = create_componentsets_in_HERON("/fake/folder", "/fake/heron_input.xml")
 
-    # Add more tests here to cover other conditions and edge cases
+    # Verify open function was called on correct files
+    for file in files_list:
+      # if file should have been opened
+      if file in ['componentSet.json', 'componentSetStuff.txt']:
```
Comment (GabrielSoto-INL) [inline tests/unit_tests/test_heron.py:656]: might be good to have a list before the `for` loop which has the acceptable filenames (e.g., `acceptable_files`) then refer to it in this line. That way it is easier to modify should we ever need to

**Review by GabrielSoto-INL — CHANGES_REQUESTED** [review]
Great work, VERY thorough addition to the FORCE testing suite! I have added a couple of comments and suggestions to improve the code so far, and had just a couple of questions.  

I think overall, would try to consolidate some repeated lines of code wherever possible.

File: tests/unit_tests/test_heron.py [hunk 5]
```diff
@@ -1,61 +1,669 @@
...
+          self.assertIn('npv_exempt', cashflow[0].keys())
+        
+        # Children
+        with self.subTest("CashFlow children nodes were not updated correctly"):
+          cf_contents = [e for e in cashflow[0].findall('./')
+                         if not e.tag is ET.Comment] # Filters out ET.Comment elements
+          self.assertEqual(len(cf_contents), 4)
+
+          # Reference driver
+          ref_driver = cashflow[0].findall('./reference_driver')
```
[the remark this answers is not in the document]
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:336]: I think that could be really helpful for simplifying the tests and also improving the clarity of the results. I'll set up some methods/functions for grouping these frequently used parse-and-assert combinations.

File: tests/unit_tests/test_heron.py [hunk 6]
```diff
@@ -1,61 +1,669 @@
...
+    cashflows = economics.findall('CashFlow')
+    self.assertEqual(len(cashflows), 1)
+    self.assertEqual(cashflows[0].attrib['name'], 'NewComponent_capex')
+    self.assertIn('type', cashflows[0].keys())
+    self.assertIn('taxable', cashflows[0].keys())
+    self.assertIn('inflation', cashflows[0].keys())
+    self.assertIn('mult_target', cashflows[0].keys())
+    
+    # Verify reference price
+    ref_price_value = cashflows[0].find('./reference_price/fixed_value')
```
[the remark this answers is not in the document]
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:62]: This should be addressed by the new methods being added. In either case, `find()` returns None if no matches are found; thus, the error is caught, albeit without a great deal of clarity as to the cause of the issue in this version.

File: tests/unit_tests/test_heron.py [hunk 7]
```diff
@@ -1,61 +1,669 @@
...
+      cases = result_tree.findall('./Case')
+      self.assertEqual(len(cases), 1)
+      self.assertIsNotNone(cases[0].find('./untouched_content_Case'))
+
+    # Verify Component nodes were merged
+    component_nodes = result_tree.findall("./Components/Component")
+    self.assertEqual(len(component_nodes), 3)
+
+    for comp in component_nodes:
+      if comp.get('name') == 'ExistingComponent0':
```
[the remark this answers is not in the document]
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:133]: A `match-case` would certainly be stylistically preferable. Is it acceptable to put it in though if it's only valid with python 3.10 and above? I'm not sure with what versions of python FORCE is intended to be compatible.

File: tests/unit_tests/test_heron.py [hunk 8]
```diff
@@ -1,61 +1,669 @@
...
+    # Set up the parse mock to return an XML tree
+    mock_parse.return_value = self.tree
+    # No additional data for XML
+    mock_listdir.return_value = ["componentSetFake.json"]
+
+    # Call the function
+    result_tree = create_componentsets_in_HERON("/fake/folder", "/fake/heron_input.xml")
+
+    # Verify Case node was transferred
+    with (self.subTest("Case node has been corrupted")):
```
[the remark this answers is not in the document]
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:123]: No, there is no reason for this. The extra parentheses have been removed

File: tests/unit_tests/test_heron.py [hunk 9]
```diff
@@ -1,61 +1,669 @@
...
+
+    for comp in component_nodes:
+      if comp.get('name') == 'Component0':
+        economics = comp.find('./economics')
+        
+        # Verify non-capex cashflow was not corrupted
+        with self.subTest("Non-capex CashFlow node was corrupted"):
+          cashflow_non_capex = economics.findall('./CashFlow[@name="other"]')
+          self.assertEqual(len(cashflow_non_capex), 1)
+          cf_noncap_contents = [e for e in cashflow_non_capex[0].findall('./')
```
[the remark this answers is not in the document]
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:277]: Yes, the purpose of this (lines 277-278) is to filter the children of the `CashFlow[@name="other"]` node to find all non-comment subelements (i.e., subnodes). This filtering is necessary because the length of the resulting list is checked. A `getchildren()` method would be cleaner than `findall('./')`, but there is not one included in the `ET` package. The `iter()` or `iterfind()` methods could be used in place of the `findall()` method, but neither would simplify syntax; xml pathing (the `'./'` argument) would still be required. Documentation here: https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.Element

File: tests/unit_tests/test_heron.py [hunk 10]
```diff
@@ -1,61 +1,669 @@
...
+        
+      elif comp.get('name') == 'Component1':
+        economics = comp.findall('./economics')
+        self.assertEqual(len(economics), 1)
+
+        # Verify number of cashflows
+        self.assertEqual(len(economics[0].findall('./CashFlow')), 2)
+
+        # Verify non-capex cashflow was not corrupted
+        with self.subTest("Non-capex CashFlow node was corrupted"):
```
[the remark this answers is not in the document]
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:295]: This non-capex cashflow differs from the one in `Component0` in that this has a `type` attribute and no subnodes (see line 280 in `Component0` and line 298 in `Component1`). They could be combined, but for the sake of clarity, I think they're better separate.

File: tests/unit_tests/test_heron.py [hunk 11]
```diff
@@ -1,61 +1,669 @@
...
+                                <CashFlow name="other" type="one-time">
+                                </CashFlow>
+                                <CashFlow name="capex" type="one-time">
+                                </CashFlow>
+                              </economics>
+                            </Component>
+
+                            <Component name="Component2">
+                              <economics>
+                                <ProjectTime>1</ProjectTime>
```
[the remark this answers is not in the document]
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:182]: Yes, `lifetime` would be correct for a real HERON input, so that's a much better example of an `economics` subnode that should not be modified.

File: tests/unit_tests/test_heron.py (hunk 2 above)
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:190]: The reason I chose not to modify the set of ValuedParams is that, in the current functionality of the `src/heron.py` script, all the ValuedParam inputs from ASPEN HYSYS and APEA are required. The `src/heron.py` script currently trusts that the ASPEN HYSYS and APEA inputs have been provided, so it does not reliably throw an error when one is missing. Thus, neither a functional nor an erroneous response to the omission of some of the ValuedParams can be tested.

File: tests/unit_tests/test_heron.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply tests/unit_tests/test_heron.py:277]: Is this working on XML directly with ET, or in the RAVEN-defined heirarchical trees? Because we can use EDDI format trees (from WORKBENCH) as well, I would hesitate to strictly depend on the XML parsing.

File: tests/unit_tests/test_heron.py (hunk 3 above)
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:566]: This is a question whose answer could depend on some broader strategic goals for how this function will be used. A warning does sound like a safe middle-of-the-road option though. No matter the solution, I think it would be more appropriately addressed in a separate issue and PR. If functionality changes significantly, this test can also be updated then.

File: tests/unit_tests/test_heron.py [hunk 12]
```diff
@@ -1,61 +1,669 @@
...
+
+    # Call the function with patch for open function
+    with patch('builtins.open', mock_open_mult):
+      result_tree = create_componentsets_in_HERON("/fake/folder", "/fake/heron_input.xml")
+
+    # Verify Component nodes were merged
+    component_nodes = result_tree.findall("./Components/Component")
+    self.assertEqual(len(component_nodes), 4)
+
+    for comp in component_nodes:
```
[the remark this answers is not in the document]
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:269]: I've moved the `economics` check above the for loop. I think that's the only thing that's easy to consolidate. 

See my response to the comment on line 133 about the `match-case`.

File: tests/unit_tests/test_heron.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:277]: @PaulTalbot-INL This uses ET. I'm pretty confident in the functionality of the parsing for now, but it's not the easiest to work with, and the syntax can be a bit picky and hard to read. So if the EDDI format trees are significantly easier, it might be worthwhile to switch over for the sake of future edits and new tests. What's the best place to get more info on those?

[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply tests/unit_tests/test_heron.py:277]: In RAVEN utils we have the TreeStructure, which behaves like XML and mimics ET in terms of methods, but is designed to flexibly load in different heirarchical structures (XML, EDDI, HIT, etc). It may not be too hard to load the XML into the tree structure and use that instead of parsing the XML directly.

See https://github.com/idaholab/raven/blob/devel/ravenframework/utils/TreeStructure.py.

[the remark this answers is not in the document]
Comment (GabrielSoto-INL) [reply tests/unit_tests/test_heron.py:277]: just to clarify, is the suggestion here to modify the `create_componentsets_in_HERON( )` method to output a `TreeStructure` object rather than an `ElementTree` object? I think that could be valuable, though I feel like might be out of scope for this particular PR. might be worth starting a different one to change that functionality.

also wanted to note that the input to the `create_componentsets_in_HERON( )` method accepts a 'str' path to some sort of HERON input script, so it could theoretically accept either a `.xml` or potentially `.heron` (for Workbench) file which it can modify. to actually _write_ to the new file, this would involve a new integration test or modifying the existing ones...

[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply tests/unit_tests/test_heron.py:277]: I see your point. I'm fine with using XML and ET for now. We've been wanting to do something to allow InputSpec on-the-fly instantiation for dynamic component building, but we haven't figured out how to make the ends meet on that yet. This is fine with ET.

File: tests/unit_tests/test_heron.py [hunk 13]
```diff
@@ -8,7 +8,36 @@
 sys.path.append(FORCE_LOC)
 from FORCE.src.heron import create_componentsets_in_HERON
 
-class TestMinimalInput(unittest.TestCase):
+class HERONTestCase(unittest.TestCase):
+  
+  def check_reference_price(self, cashflow, correct_value, correct_content_length=1):
```
Comment (GabrielSoto-INL) [inline tests/unit_tests/test_heron.py:13]: would you be able to add docstrings to these new methods? For reference: https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#function-and-class-documentation-strings

File: tests/unit_tests/test_heron.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (GabrielSoto-INL) [reply tests/unit_tests/test_heron.py:158]: on second thought, based on your note, should be fine as is

File: tests/unit_tests/test_heron.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (GabrielSoto-INL) [reply tests/unit_tests/test_heron.py:300]: @caleb-sitton-inl seems like it's best to keep as is, but also it might be a good idea to start an issue for Feature Request regarding switching to dynamic `TreeStructure` elements rather than ET

**Review by GabrielSoto-INL — CHANGES_REQUESTED** [review]
Sorry, one more request: could you add docstrings to the new methods at the top of the HERONTestCase class? otherwise, looks good!

File: tests/unit_tests/test_heron.py (hunk 13 above)
Comment (caleb-sitton-inl) [reply tests/unit_tests/test_heron.py:16]: No problem.

**Review by GabrielSoto-INL — APPROVED** [review]
Changes LGTM, great work!
